GBPTree writer fail fast if hitting tree node with valid successor

burqen opened the pull request:
If writer reach a tree node with a valid successor it means that the
tree is not consistent in one out of three ways:
1. Tree state point to a root that is outdated (has a successor)
2. Child pointer in parent does not point to latest successor of child.
3. Sibling pointer does not point to latest successor of sibling.

Commit:
GBPTree writer fail fast if hitting tree node with valid successor

If writer reach a tree node with a valid successor it means that the
tree is not consistent in one out of three ways:
1. Tree state point to a root that is outdated (has a successor)
2. Child pointer in parent does not point to latest successor of child.
3. Sibling pointer does not point to latest successor of sibling.

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTree.java:
```diff
@@ -987,8 +1002,15 @@ public void put( KEY key, VALUE value ) throws IOException
+            }
+            catch ( TreeInconsistencyException e )
+            {
+                throw appendTreeInformation( e );
```
why do we not close cursor in this case?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTree.java:
```diff
@@ -1024,8 +1046,17 @@ private void setRoot( long rootPointer )
+            }
+            catch ( TreeInconsistencyException e )
+            {
+                throw appendTreeInformation( e );
```
and here?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/PointerChecking.java:
```diff
@@ -26,6 +26,12 @@
+    static final String WRITER_TRAVERSE_OLD_STATE_MESSAGE =
+                    "Writer traversed to a tree node that has a valid successor, " +
+                    "meaning writer will not act on latest version of the tree. This is not allowed. " +
+                    "This is most likely due to failure to checkpoint the tree before shutdown and/or tree state " +
```
also as we discuss, message ask something that we do not want our users to do

MishaDemianenko reviewed (commented):
Couple of comments

Commit:
Cleaner lifecycle of GBPTree.Writer

Initialize SingleWriter now always leave writer in a fully
consistent state, meaning no additional cleanup is
necessary from the outside. In all cases when writer is
not successfully initialized, some exception will be thrown.

Commit:
Don't encourage users to delete store files

burqen commented:
@MishaDemianenko Comments addressed 🎉

Commit:
Checkstyle
